docs/source/testing.rst

```diff
@@ -0,0 +1,316 @@
+
+Ex-ante: Defining categories of tests
+*************************************
+
+Another option to subselect a certain set of tests is by use of
+`pytest markers <https://docs.pytest.org/en/7.1.x/example/markers.html>`_.
+The following is one way of using markers in conjunction with ``datajudge``.
+
+In this particular illustration we'll allow for two markers:
+
+* ``basic``: indicating that only truly fundamental tests should be run
+* ``all``: indicating that any available test should be run
+
+For that matter we'll add a bit of pytest magic to the respective ``conftest.py``.
+
+.. code-block:: python
+  :caption: ``conftest.py``
+
+  def pytest_generate_tests(metafunc):
+      if "basic_constraint" in metafunc.fixturenames:
+           metafunc.parametrize(
+	       "basic_constraint",
+               metafunc.module.get_basic_constraints(),
+               ids=metafunc.module.idfn,
+	   )
+      if "constraint" in metafunc.fixturenames:
+          metafunc.parametrize(
+	      "constraint",
+	      metafunc.module.get_all_constraints(),
+	      ids=metafunc.module.idfn,
+          )
+
+
+Moreover, we'll have to register these markers in pytest's ``pytest.ini`` file.
+You can read more about these files
+`here <https://docs.pytest.org/en/6.2.x/customize.html>`_.
+
+
+.. code-block::
+   :caption: ``pytest.ini``
+
+   [pytest]
+   addopts = --strict-markers
+   markers = basic: basic specification
+	     all: entire specification
+
+Once that is taken care of, one can adapt one's specification as follows:
+
+.. code-block:: python
+  :caption: ``specification.py``
+
+  def get_basic_requirements() -> List[Requirement]:
+      ...
+
+  def get_advanced_requirements() -> List[Requirement]:
+      ...
+
+  def get_basic_constraints() -> List[Constraint]:
+      return [constraint for requirement in get_basic_requirements() for constraint in requirement]
+
+  def get_all_constraints() -> List[Constraint]:
+      all_requirements = get_basic_requirements() + get_advanced_requirements()
+      return [constraint for requirement in all_requirements for constraint in requirement]
+
+  @pytest.mark.basic
+  def test_basic_constraint(basic_constraint: Constraint, datajudge_engine):
+      test_result = basic_constraint.test(datajudge_engine)
+      assert test_result.outcome, test_result.failure_message
+
+  @pytest.mark.all
+  def test_all_constraint(constraint: Constraint, datajudge_engine):
+      test_result = constraint.test(datajudge_engine)
+      assert test_result.outcome, test_result.failure_message
+
+Once these changes are taken care of, one may run
+
+.. code-block:: console
+
+  $ pytest specification.py -m basic
+
+to only test the basic ``Requirement`` s or
+
+.. code-block:: console
+
+  $ pytest specification.py -m all
+
+to test all ``Requirement`` s.
+
+
+Parametrization
+---------------
+
+A given specification might rely on identifiers such as database names or
```

Comment:
I find the two first paragraphs a bit too abstract. And while the first one might be fine as an opening one, the second doesn't help much in supporting and making it more concrete.

Once the example comes, it's more understandable, but I also am familiar with the intention.

Reply:
Though I easily understood them on the second read, I'd agree.
Maybe something more like 

---
A given _test_ specification might rely on identifiers such as database names or table names. Moreover it might be that, e.g. when iterating from one _version_ of the data to another, these _names_ change. ~All the while the expectations towards how the data relates to another might remain.~

In other words, it could be that the test logic ~logic of the specification~ should remain unchanged while its pointers to data might change. Therefore, one ~One might simply rely on hard-coded references to data. Yet, one~ might just as well parametrize such pointers or identifiers.

---
Or a much shorter version might also suffice:

> For applying the same test suite to different data sources and targets, e.g. different versions over time, one might want to parameterize pointers such as database names or table identifiers.

Reply:
Somethign I forgot to mention is that `parametrization` here clashes with what a `parametrize` mark in `pytest`.

I think you can safely use the word `parameter`, but not `parametrization` which might cause some confusion at first. To me, what it's happening here evokes the concept of `templating` in a way.

Reply:
Thanks a lot for your thoughts, both of you. I tried to distill them into an altered version of this section.